### DISCUSSION TITLE
Setting touch detection threshold

### DISCUSSION THREAD

**Discussion by rogerknop** [Bodmer/TFT_eSPI#2461] [Alerts]
I really like this library, but currently I am struggeling with the touch functionality.

My board [ILI9488](https://www.amazon.de/dp/B08GBYQQV5?psc=1&ref=ppx_yo2ov_dt_b_product_details)
My [ESP DevKitC](https://www.amazon.de/dp/B0811KLGDD?psc=1&ref=ppx_yo2ov_dt_b_product_details)
[Wiring](https://cdn.hackaday.io/images/4265851608817059473.png)

Setup: [Setup204_ESP32_TouchDown.h](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup204_ESP32_TouchDown.h)

The problem is that I cannot calibrate the TFT. It is doing that without touching it on its own.
I cannot influence that. The self-calibating is too fast :-)

Now I have tested with Test_Touch_Controller.ino and the output of z is changing all the time without touching:
```
x: 0     y: 4095     z: 481 
x: 0     y: 4095     z: 481 
x: 0     y: 4095     z: 479 
x: 0     y: 4095     z: 483 
x: 0     y: 4095     z: 487 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 481 
x: 0     y: 4095     z: 481 
x: 0     y: 4095     z: 471 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 508 
x: 0     y: 4095     z: 486 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 489 
x: 0     y: 4095     z: 487 
x: 0     y: 4095     z: 490 
x: 0     y: 4095     z: 486 
x: 0     y: 4095     z: 489 
x: 0     y: 4095     z: 459 
x: 0     y: 4095     z: 489 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 492 
x: 0     y: 4095     z: 481 
x: 0     y: 4095     z: 486 
x: 0     y: 4095     z: 490 
x: 0     y: 4095     z: 482 
x: 0     y: 4095     z: 483 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 489 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 485 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 491 
x: 0     y: 4095     z: 490 
x: 0     y: 4095     z: 486 
x: 0     y: 4095     z: 484 
x: 0     y: 4095     z: 486 
x: 0     y: 4095     z: 490 
x: 0     y: 4095     z: 484 
```

If I tab the values are changing - so in general it seems to work.

Any help would be appreciated!
Thanks in advance, Roger

**Comment by rogerknop** [comment]
Read_User_Setup.ino Output:
```
TFT_eSPI ver = 2.5.23
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9488
Display width  = 320
Display height = 480

MOSI    = GPIO 23
MISO    = GPIO 19
SCK     = GPIO 18
TFT_CS   = GPIO 15
TFT_DC   = GPIO 2
TFT_RST  = GPIO 4
TOUCH_CS = GPIO 21

TFT_BL           = GPIO 32
TFT_BACKLIGHT_ON = HIGH

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
Touch SPI frequency   = 2.50
```

**Comment by rogerknop** [comment]
I had another discussion about that issue in the [forum](https://forum.arduino.cc/t/touch-is-not-working/1096318/12)

Here the workaround:
Most of the examples are using the function tft.validTouch()
This function has a threshold for z.
In this function in the library I changed the threshold value fix to 500.
Now it works!

Is it planned to control that via setup? From coding perspective it does not look like.

**Comment by Bodmer** [comment]
The threshold can be set in the getTouch call, for example here the Z pressure threshold is set to 600:
```
uint16_t touchX = 0, touchY = 0;
bool touched = tft.getTouch( &touchX, &touchY, 600 );
```
Note that for an ILI9488 screen the display SDO pin must be left unconnected.

**Comment by rogerknop** [comment]
getTouch with theshold is fine, but the function calibrateTouch in Touch.cpp is using the constant Z_THRESHOLD and this function is even deviding the value by 2? Why is that devided?
Probably the calibrateTouch also needs the optional parameter theshold.

**Reply by Bodmer** [reply]
As per comment in code the sensitivity in corners is much lower. It seems the touch screen you have is not behaving normally.

**Comment by rogerknop** [comment]
Yes understood. But I thought you have added the threshold parameter for exactly such a use case for "unnormal" touch screens, so that it is not necessary to modify your coding.
Therefore I was hoping that you could also enhance the calibrateTouch by this parameter for a fully workaround support.

**Comment by Bodmer** [comment]
With the latest 2.5.23 version you can add the threshold to the setup, eg:
#define Z_THRESHOLD 1200

**Comment by rogerknop** [comment]
But this is only for calibrateTouch and will be devided by 2.
For getTouch a parameter is needed in addition.
Correct?

**Reply by Bodmer** [reply]
Yes, I may update the library to try and auto set the threshold during calibration as it seems to be particularly high on some boards.
